extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -0,0 +1,1263 @@
+  private final IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+  private final KafkaIndexTaskClient taskClient;
+  private final ObjectMapper sortingMapper;
+  private final KafkaSupervisorSpec spec;
+  private final String dataSource;
+  private final KafkaSupervisorIOConfig ioConfig;
+  private final String supervisorId;
+
+  private final ExecutorService exec;
+  private final BlockingQueue<Notice> notices = new LinkedBlockingDeque<>();
+  private final Timer timer = new Timer(true);
+  private final Object stopLock = new Object();
+
+  private KafkaConsumer consumer;
+  private DateTime firstRunTime;
+  private boolean listenerRegistered = false;
+  private long lastRunTime;
+
+  private volatile boolean started = false;
+  private volatile boolean stopped = false;
+
+  public KafkaSupervisor(
+      TaskStorage taskStorage,
+      TaskMaster taskMaster,
+      IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator,
+      KafkaIndexTaskClient taskClient,
+      ObjectMapper mapper,
+      KafkaSupervisorSpec spec
+  )
+  {
+    this.taskStorage = taskStorage;
+    this.taskMaster = taskMaster;
+    this.indexerMetadataStorageCoordinator = indexerMetadataStorageCoordinator;
+    this.taskClient = taskClient;
+    this.sortingMapper = mapper.copy().configure(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY, true);
+    this.spec = spec;
+
+    this.dataSource = spec.getDataSchema().getDataSource();
+    this.ioConfig = spec.getIoConfig();
+    this.supervisorId = String.format("KafkaSupervisor-%s", dataSource);
+    this.exec = Execs.singleThreaded(supervisorId);
+  }
+
+  @Override
+  public void start()
+  {
+    Preconditions.checkState(!started, "already started");
+    Preconditions.checkState(!exec.isShutdown(), "already stopped");
+
+    try {
+      consumer = getKafkaConsumer();
+
+      exec.submit(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              try {
+                while (!Thread.currentThread().isInterrupted()) {
+                  final Notice notice = notices.take();
+
+                  try {
+                    notice.handle();
+                  }
+                  catch (Exception e) {
+                    log.makeAlert(e, "KafkaSupervisor[%s] failed to handle notice", dataSource)
+                       .addData("noticeClass", notice.getClass().getSimpleName())
+                       .emit();
+                  }
+                }
+              }
+              catch (InterruptedException e) {
+                log.info("KafkaSupervisor[%s] interrupted, exiting", dataSource);
+              }
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Exception starting KafkaSupervisor[%s]", dataSource)
+         .emit();
+      throw Throwables.propagate(e);
+    }
+
+    firstRunTime = DateTime.now().plus(ioConfig.getStartDelay());
+    timer.scheduleAtFixedRate(
```

Comment:
most other Druid code uses `Execs.scheduledSingleThreaded(threadName)` rather than a timer. That makes sure we always get a daemon thread with a nice name. Looks like you already have it using a daemon thread, but a name would be nice too.

So I would either switch this to a scheduled executor (with a nice name) or use the Timer constructor that takes a thread name.